### ISSUE TITLE
Only number index type is inferred from `any` computed property key

### ISSUE THREAD

**Issue by vepanimas** [microsoft/TypeScript#45629]
# Bug Report

### 🔎 Search Terms

* Computed properties
* Indexers
* Index signatures

### 🕗 Version & Regression Information

- This is the behavior in every version I tried.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.0-dev.20210826#code/CYUwxgNghgTiAEEQBd5QFxoHYE8DcAUAWAPZYDOqJARgFbwC88A3gfPANpQC6mAROT4AaAgF888APSSWnAB6YsAVwC21EDF7xKMAJZYA5vFFEatDnyjUwfbniA)

### 💻 Code

```ts
declare let a: any;

const obj = {
  [a]: "s",
}; // { [x: number]: string }

obj["abc"]; // error
```

### 🙁 Actual behavior

A numeric indexer is inferred. String index access marked as invalid.

### 🙂 Expected behavior

A string indexer is expected. No error for string access.

```ts
declare let a: any;

const obj = {
  [a]: "s",
}; // { [x: number]: string }

obj["abc"]; // no error
```

I created a small PR that addresses the issue as an example but I'm still not sure if the current behavior is intentional or not.

EDIT: I'm still not sure if the inferred indexer should have a String type, perhaps a union of the allowed primitive types fits better, e.g.:
```ts
declare let a: any;

const obj = {
  [a]: "s",
}; // { [x: string | number | symbol]: string }
```

**Comment by andrewbranch** [comment]
@weswigham was aware of this bug but I can’t find an earlier instance of it logged. He did mention the fix could be breaky, so no guarantees we can actually take a PR, but we can take a look.

**Comment by vepanimas** [comment]
@andrewbranch agree, replacing a numeric to a string indexer, in theory, can break something, but adding a `{ [x: string | number | symbol]: string }` instead shouldn't AFAIU. Maybe it would be a better solution.
